tracer/src/Datadog.Trace/Debugger/Caching/DefaultMemoryChecker.cs

```diff
@@ -0,0 +1,166 @@
+    [DllImport("kernel32.dll", CharSet = CharSet.Auto, SetLastError = true)]
+    private static extern bool GlobalMemoryStatusEx([In, Out] MEMORYSTATUSEX lpBuffer);
+
+    public bool IsLowResourceEnvironment()
+    {
+        return _isLowResourceEnvironment;
+    }
+
+    private bool CheckLowResourceEnvironment()
+    {
+        try
+        {
+            Logger.Information("Checking if environment is low on resources");
+            // Check if we're using more than 75% of available memory or there is less than 1GB of RAM available.
+            return IsLowResourceEnvironmentGc() || IsLowResourceEnvironmentSystem();
+        }
+        catch (Exception ex)
+        {
+            Logger.Error(ex, "Failed to check if environment is low on resources. Assuming it is not.");
+            return false;
+        }
+    }
+
+    private bool IsLowResourceEnvironmentGc()
+    {
+#if NETCOREAPP3_0_OR_GREATER
+
+        long totalMemory = GC.GetTotalMemory(false);
+        long totalAvailableMemory = GC.GetGCMemoryInfo().TotalAvailableMemoryBytes;
+
+        // Check if we're using more than 75% of available memory.
+        return totalMemory > (totalAvailableMemory * 0.75);
+#else
+        return false;
+#endif
+    }
+
+    private bool IsLowResourceEnvironmentSystem()
+    {
+        return FrameworkDescription.Instance.IsWindows() ? CheckWindowsMemory() : CheckUnixMemory();
+    }
+
+    internal bool CheckWindowsMemory()
+    {
+        try
+        {
+            if (MEMORYSTATUSEX.GetAvailablePhysicalMemory(out var availableMemory))
+            {
+                // If less than 1GB of RAM is available, consider it a low-resource environment
+                return availableMemory < 1_073_741_824; // 1 GB in bytes
+            }
+        }
+        catch (Exception e)
+        {
+            Logger.Error(e, "Fail to call GlobalMemoryStatusEx");
+
+            // If we can't access memory info, we'll fall back to the default capacity
+            throw;
+        }
+
+        return false;
+    }
+
+    internal bool CheckUnixMemory()
+    {
+        try
+        {
+            // for linux we can check /proc/meminfo
+            var memAvailable = ReadMemInfo();
+            if (long.TryParse(memAvailable, out long availableKB))
+            {
+                return availableKB * 1024 < LowMemoryThreshold;
+            }
+        }
+        catch (Exception e)
+        {
+            Logger.Error(e, "Fail to read or parse /proc/meminfo");
+
+            // If we can't read memory info, we'll fall back to the default capacity
+            throw;
+        }
+
+        return false;
+    }
+
+    protected virtual string? ReadMemInfo()
+    {
+        string memInfo = System.IO.File.ReadAllText("/proc/meminfo");
+        var memAvailable = memInfo.Split('\n')
+                                  .FirstOrDefault(l => l.StartsWith("MemAvailable:"))
+                                 ?.Split(':')[1].Trim().Split(' ')[0];
```

Comment:
Not critical since this is normally only called once, but we could still try to not use `String.Split()` or `String.Trim()` to avoid allocating multiple strings and arrays. You can use the `string.SplitIntoSpans()` extension method for this. See `W3CBaggagePropagator.ParseHeader()` for an example usage.

Once you have the final `Span<char>`, you will have to convert it to string in older target frameworks to use `long.TryParse()`, but in more recent TFMs you can parse the `Span<char>` directly, too.

Reply:
See updated implementation in 93ce326e53379d4935cf896c25e5600ea89b4164 and fa9454fbb827e4641a220ce9b273d4f915a77814

Reply:
Nice, thanks!